zenoh/src/api/builders/close.rs

```diff
@@ -0,0 +1,151 @@
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use std::{
+    future::{Future, IntoFuture},
+    pin::Pin,
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use zenoh_core::{Resolvable, Wait};
+use zenoh_result::ZResult;
+use zenoh_runtime::ZRuntime;
+
+/// A builder for close operations.
+// NOTE: `Closeable` is only pub(crate) because it is zenoh-internal trait, so we don't
+// care about the `private_bounds` lint in this particular case.
+#[allow(private_bounds)]
+pub struct CloseBuilder<TCloseable: Closeable> {
+    closee: TCloseable::TClosee,
+    timeout: Duration,
+}
+
+// NOTE: `Closeable` is only pub(crate) because it is zenoh-internal trait, so we don't
+// care about the `private_bounds` lint in this particular case.
+#[allow(private_bounds)]
+impl<TCloseable: Closeable> CloseBuilder<TCloseable> {
+    pub(crate) fn new(closeable: &'_ TCloseable) -> Self {
+        Self {
+            closee: closeable.get_closee(),
+            timeout: Duration::from_secs(10),
+        }
+    }
+
+    #[cfg(all(feature = "unstable", feature = "internal"))]
+    /// Set the timeout for close operation
+    ///
+    /// # Arguments
+    ///
+    /// * `timeout` - The timeout value for close operation (10s by default)
+    ///
+    #[zenoh_macros::unstable_doc]
+    pub fn timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = timeout;
+        self
+    }
+
+    #[cfg(all(feature = "unstable", feature = "internal"))]
+    /// Run Close operation concurrently
+    #[zenoh_macros::unstable_doc]
+    pub fn in_background(
+        self,
+    ) -> BackgroundCloseBuilder<<CloseBuilder<TCloseable> as Resolvable>::To> {
+        BackgroundCloseBuilder::new(self.into_future())
+    }
+}
+
+impl<TCloseable: Closeable> Resolvable for CloseBuilder<TCloseable> {
+    type To = ZResult<()>;
+}
+
+impl<TCloseable: Closeable> Wait for CloseBuilder<TCloseable> {
+    fn wait(self) -> Self::To {
+        ZRuntime::Application.block_in_place(self.into_future())
+    }
+}
+
+impl<TCloseable: Closeable> IntoFuture for CloseBuilder<TCloseable> {
+    type Output = <Self as Resolvable>::To;
+    type IntoFuture = Pin<Box<dyn Future<Output = <Self as IntoFuture>::Output> + Send>>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        Box::pin(
+            async move {
+                if tokio::time::timeout(self.timeout, self.closee.close_inner())
+                    .await
+                    .is_err()
+                {
+                    bail!("close operation timed out!")
+                }
+                Ok(())
+            }
+            .into_future(),
+        )
+    }
+}
+
+#[cfg(all(feature = "unstable", feature = "internal"))]
+/// A builder for close operations running in background
+// NOTE: `Closeable` is only pub(crate) because it is zenoh-internal trait, so we don't
+// care about the `private_bounds` lint in this particular case.
+#[zenoh_macros::unstable_doc]
```

Comment:
For internal data it should actually be `[doc(hidden)]`.

Reply:
fixed